clientconn.go

```diff
@@ -476,6 +478,97 @@ func (s ConnectivityState) String() string {
 	}
 }
 
+// connectivityStateEvaluator gets updated by addrConns when their
+// states transition, based on which it evaluates the state of
+// ClientConn.
+// Note: This code will eventually sit in the blanacer in the new design.
+type connectivityStateEvaluator struct {
+	csMgr               *connectivityStateManager
+	mu                  sync.Mutex
+	numReady            uint64 // Number of addrConns in ready state.
+	numConnecting       uint64 // Number of addrConns in connecting state.
+	numTransientFailure uint64 // Number of addrConns in transientFailure.
+}
+
+// recordTransition records state change happening in every addrConn and based on
+// that it evaluates what state the ClientConn is in.
+// It can only transition between Ready, Connecting and TransientFailure. Other states,
+// Idle and Shutdown are transitioned into by ClientConn; in the begining of the conneciton
+// before any addrConn is created ClientConn is in idle state. In the end when ClientConn
+// closes it is in Shutdown state.
+// TODO Note that in later releases, a ClientConn with no activity will be put into an Idle state.
+func (cse *connectivityStateEvaluator) recordTransition(oldState, newState ConnectivityState) {
+	cse.mu.Lock()
+	defer cse.mu.Unlock()
+
+	// Update counters.
+	for idx, state := range []ConnectivityState{oldState, newState} {
+		updateVal := 2*uint64(idx) - 1 // -1 for oldState and +1 for new.
+		switch state {
+		case Ready:
+			cse.numReady += updateVal
+		case Connecting:
+			cse.numConnecting += updateVal
+		case TransientFailure:
+			cse.numTransientFailure += updateVal
+		}
+	}
+
+	// Evaluate.
+	if cse.numReady > 0 {
+		cse.csMgr.updateState(Ready)
+		return
+	}
+	if cse.numConnecting > 0 {
+		cse.csMgr.updateState(Connecting)
+		return
+	}
+	cse.csMgr.updateState(TransientFailure)
+}
+
+// connectivityStateManager keeps the ConnectivityState of ClientConn.
+// This struct will eventually be exported so the balancers can access it.
+type connectivityStateManager struct {
+	mu         sync.Mutex
+	state      ConnectivityState
+	notifyChan chan struct{}
+}
+
+// updateState updates the ConnectivityState of ClientConn.
+// If there's a change it notifies goroutines waiting on state change to
```

Comment:
Although rare, but it's possible multiple goroutines can make a call to WaitForStateChange